enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,314 @@
+- The new Dnsmasq service should be responsible to resolve 'api-int' (to API-VIP value provided in install-config) hostname when CoreDNS-MDNS pod isn't running on the node.
+- CoreDNS-MDNS should register himself as an upstream server in dnsmasq [via dbus interface](https://github.com/imp/dnsmasq/blob/master/dbus/DBus-interface#L51) for ClusterName.DomainName
+ 
+#### Self hosted API Loadbalancer
+- The self hosted API loadbalancer will run similarly to the current mode with one difference: HAProxy pods will run by CHSNO as DaemonSet pods instead of the static pods.
+- API should be available even when HAProxy pods don't run.
+- Additional functionality aspects of the self-hosted Loadbalancer maybe further analyzed upon need in a separate enhancement. 
+
+#### Master node shutdown
+
+In order not to break the [graceful termination process of kube-api](https://github.com/openshift/installer/blob/master/docs/dev/kube-apiserver-health-check.md) the pods responsible for API reachability should be deleted in the following order: kube-api, Keepalived, and HAProxy.
+
+Using [K8S graceful node shutdown in kubelet](https://github.com/kubernetes/kubernetes/pull/96129) it should be possible to delete the pods in the order described above.
+
+A detailed design of pods priority will be defined after graceful node shutdown in kubelet feature will be integrated in Openshift.
+
+Question: Isn't it possible to force deletion of kube-api, Keepalived, and HAProxy in the desired order by implementing smart delay code in SIGTERM handler in each one of the pods?
+
+#### Disaster recovery support
+
+In order to support disaster recovery kubelet, kube-controller-manager and kube-scheduler should be able to communicate with kube-api even when all CHNSO pods are shutdown.
+
+The following steps will ensure that kube-api is available post disaster recovery:
+- Dnsmasq service and Keepalived static pods will start running on master nodes
+- API-VIP address will be assigned to one of the master nodes by Keepalived
+- Dnsmasq should resolve api-int hostname to API-VIP address
+
+The above steps should be applied also for cluster shutdown and resume use-case.
+
+#### Timelined events: node shutdown, node restart, disaster recovery
+As the CHNSO allows admins to use their own external Load Balancing and DNS resolution instead of the self-hosted stack, this section illustrates the events (in chronological order) for both self-hosted and external LB cases.
+
+##### Shutdown of master node (x) holds the VIP - self-hosted LB
+- T0: kube-apiserver reports failure on /readyz
+- T1: (<= T0 + 2Sec) HAProxy will mark master-x backend as failed, and should send new connections towards nodes != master-x.
+- T2: (= T0 + kube-api graceful termination timeout=90Sec ) kube-apiserver,Keepalived and HAProxy pods shutdown, Keepalived sends a priority 0 message to trigger VIP switchover.
```

Comment:
please add a note why 90s.

Reply:
not sure you discussed that elsewhere already, but: this setup will receive requests between T2-60 and T2, forward them to another master and cuts them off when shutting down.

This is like that today as static pod haproxy. But be clear here. It's not a blocker because it is not a regression.